### PR TITLE
add machine images info to labels

### PR DESCRIPTION
**Pull request by tedteng**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement
/platform gcp

**What this PR does / why we need it**:
add os information as labels in machine class objects
image

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/machine-controller-manager/issues/880

**Special notes for your reviewer**:

**Release note**:

```other operator
add os information as labels in machine class objects.
```
/invite @rishabh-11

### CODE REVIEW

**Comment by tedteng** [comment]
Hey, @rishabh-11  Take GCP as the first sample, if everything is all good, will open the rest of the others. (AWS,Azure,Alicloud,CCloud)

**Review by kon-angelo — CHANGES_REQUESTED** [review]

File: pkg/controller/worker/machines.go
Comment (tedteng) [reply pkg/controller/worker/machines.go]: ff28099895864cf015ab3a8013301b4f58a35e83 
btw how did you get this error return? is it from the integration test?

File: pkg/controller/worker/machines.go
Comment (elankath) [inline pkg/controller/worker/machines.go:181]: Why not move this code into function `getGceInstanceLabels` itself  ? anyways `MachineImage/Version` fields are part of `pool` obj which is passed to `getGceInstanceLabels`

**Comment by kon-angelo** [comment]
Actually after reading the issue that this PR tries to solve I am not convinced about the approach here. What is  exactly the end-goal that you are trying to achieve ? Should the annotations be in the `machine`, `machineclass` etc. or in the actual virtual machine or the `node` object?

The place you are changing now would add labels to the node object and the VM in gcp itself, which the original ticket does not mention as a solution. I would suggest that before considering this change, maybe you add a comment in the initial ticket with the exact solution you are trying to achieve, specifying all the places that you want to add these labels. We can then verify the approach first and review the subsequent PRs accordingly.

/hold

**Comment by tedteng** [comment]
> What is exactly the end-goal that you are trying to achieve? 

The main idea from our SRE colleague is to extract the os name and os version data from any one of them (machine, machineDeployment, machineclass, machinesets, node) to generate our report.  The specific location of the data, whether in labels, annotations or the spec of these objects, is not a concern.

> Should the annotations be in the machine, machineclass etc., or the actual virtual machine or the node object?

per sync with MCM colleague, The MCM is more independent and cannot directly retrieve OS name and version information from Gardener's `cloudprofile` or `worker` objects (machine image information) in the source code. https://github.com/gardener/machine-controller-manager/issues/880#issuecomment-1909798732 Consequently, we propose adding data as labels in `machineclass` object, extending the labels in `generatemachineconfig`.

Please feel free to suggest any alternative solutions if available.

File: pkg/controller/worker/machines.go
Comment (kon-angelo) [reply pkg/controller/worker/machines.go]: You get this error if you try to create a machine with invalid GCP labels. The error is from GCP API itself

**Comment by kon-angelo** [comment]
The issue does entirely lie with this PR. Let's look at the `machineclass`:

https://github.com/gardener/gardener-extension-provider-gcp/blob/6c91b945ead507f2b38d0b2aa614f2f1b64b34e0/charts/internal/machineclass/templates/machineclass.yaml#L21C1-L24C11 and https://github.com/gardener/gardener-extension-provider-gcp/blob/6c91b945ead507f2b38d0b2aa614f2f1b64b34e0/charts/internal/machineclass/templates/machineclass.yaml#L43C1-L46C11

You see the labels are used both in the k8s object as well as the VM creation call. Now you try to sanitize the labels to adhere to the restrictions of GCP API, but the OS image name will not match what is in the cloudprofile (due to the transformation). 

Different providers have different standards regarding the allowed characters. In the end the SREs trying to retrieve the information of the OS, would still need to know how the provider does the transformation to valid labels to match the name as it is in the cloudprofile.

Maybe we should split the labels for the machineclass with the labels that we use for the machine itself. In provider-azure for example (https://github.com/gardener/gardener-extension-provider-azure/blob/master/charts/internal/machineclass/templates/machineclass.yaml) we don't use any labels for the machineclass itself. I don't think that the labels as they are offer much value as the information are basically duplicated inside it's spec in the machine. We could therefore either decide to use a separate set of labels for the machineclass or use them unsanizited.

**Comment by tedteng** [comment]
push 4e7b3c81f8466ddb3eafc29d55c174180bc336be, The testing process is to edit `machineclass` object in the real dev cluster to add two "labels", then delete the `machine` object and `node` object, and then waiting both objects recreated. no issue so far .

**Review by kon-angelo — CHANGES_REQUESTED** [review]
These are not labels of the machineclass. These are properties of the machine class's provider spec as defined here: https://github.com/gardener/machine-controller-manager-provider-gcp/blob/8e6f154abbace7c6fb4a85051191de8211739677/pkg/api/v1alpha1/provider_spec.go#L207

This just adds labels on the provider side for disks but it doesnt change anything on the k8s side

**Comment by rishabh-11** [comment]
>Maybe we should split the labels for the machineclass with the labels that we use for the machine itself

Yes, we should do this.

I have one question. I looked at the code and am unable to understand how some of the fields go under `providerSpec` for the machine class and the rest do not. And how are the labels for the machine class and `providerSpec` in the machine class kept the same? Is it applied in a way where it looks for the fields and simply replaces them when found and does not care about anything else?

**Comment by tedteng** [comment]
The default label managed in https://github.com/gardener/machine-controller-manager/blob/7f7878e3176f495a1d8fb031f0105486f0cc2937/pkg/apis/machine/v1alpha1/machineclass_types.go#L41, it also passed to GCP API call.
if I add a new field called `machineClassLables`  then I need to extend a new property in https://github.com/gardener/machine-controller-manager/blob/7f7878e3176f495a1d8fb031f0105486f0cc2937/pkg/apis/machine/v1alpha1/machineclass_types.go#L37C1-L46C60.  However I don't want to change the `machineclass` API type.

but after a new PR merged [ https://github.com/gardener/gardener-extension-provider-gcp/pull/669  . ](https://github.com/gardener/gardener-extension-provider-gcp/pull/669/files#diff-ede719e7797a6dd04b91acb4296b78434730159686b71e616b1c989712a7716eR191) it seems the labels are separated. I will try it add new labels in, `"labels": machineClassLabels`only and keep `poolLabels` in the rest of the places(disk, instance) to check if blocked by the GCP API call.

**Comment by tedteng** [comment]
Hi @kon-angelo  Finally I built a test environment, commit [a104548](https://github.com/gardener/gardener-extension-provider-gcp/commit/a104548d1b7274f52a2e7997f4fabec0957e049a) and tested. 
image
image
but I have no idea how to fix the `make test` in there, Do you have any idea? Thanks.

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**
